### PR TITLE
Fixes #5 and adds config for perf improvement.

### LINKED ISSUE #5: Configuration has moved to package.json instead of IDE settings
Hi, first - nice job taking over maintenance.

If I'm reading the 4.0.0 update notes correctly, `sort-imports.sort-type` has moved to `package.json` configuration instead, to be closer to the underlying [renke/import-sort](https://github.com/renke/import-sort) library.

I work in a large codebase with lots of other developers that all use different IDEs - and can't inflict my specific IDE settings on our shared package.json.

Any chance we could fall back to supporting the old VSCode `sort-imports.sort-type` config if values aren't specified in package.json?

### PR DESCRIPTION
**Pull request by cliffkoh**
* Introduced `sort-imports.default-sort-style`, which defaults to `module-compact`. Other possible values are `module` and `eslint`.
* Introduced `sort-imports.cache-package-json-config-checks` which defaults to `true`. When true, will cache calls to `import-sort-config` thereby improving performance 
  (avoids repeated non-trival disk lookups and parsing).
* Fix bug in `Save file without sorting import` which caused it to not work.

### CODE REVIEW

File: src/sort.ts [hunk 1]
```diff
@@ -26,11 +30,27 @@ export function sort(document: TextDocument): string {
...
+        if (useCache && cachedParser === undefined) {
+            const { parser, style } = getConfig(extension, directory, config);
+            cachedParser = parser;
+            cachedStyle = style;
+            const result = importSort(currentText, parser, style, fileName);
+            return result.code;
+        } else {
+            const result = importSort(currentText, cachedParser, cachedStyle, fileName);
+            return result.code;
+        }
```
Comment (amatiasq) [inline src/sort.ts]: Can we change the `if`/`else` to remove duplication?

```javascript
if (useCache && !cachedParser) {
    const { parser, style } = getConfig(extension, directory, config);
    cachedParser = parser;
    cachedStyle = style;
}

const result = importSort(currentText, cachedParser, cachedStyle, fileName);
return result.code;
```

File: package.json [hunk 2]
```diff
@@ -64,6 +64,16 @@
...
           "description": "Suppress warnings when sorting fails."
         },
+        "sort-imports.cache-package-json-config-checks": {
+          "type": "boolean",
+          "default": true,
+          "description": "If true, will cache calls to `import-sort-config`, thereby improving performance."
+        },
+        "sort-imports.default-sort-style": {
+          "type": "string",
+          "default": "module-scoped",
```
Comment (amatiasq) [inline package.json]: Default should be `eslint` otherwise we're changing the default style for all current users.

File: package.json [hunk 3]
```diff
@@ -86,12 +96,13 @@
...
-    "import-sort-style-module": "^3.0.0",
-    "import-sort-style-module-compact": "^1.0.0"
+    "import-sort": "4.2.0",
+    "import-sort-config": "^4.2.0",
+    "import-sort-parser-babylon": "^4.2.0",
+    "import-sort-parser-typescript": "^4.2.0",
+    "import-sort-style-eslint": "^4.0.0",
+    "import-sort-style-module": "^4.0.0",
+    "import-sort-style-module-compact": "^1.0.0",
+    "import-sort-style-module-scoped": "^1.0.1"
```
Comment (amatiasq) [inline package.json]: This is a nice addition, I didn't know this style. 👍

File: src/sort.ts [hunk 4]
```diff
@@ -1,16 +1,20 @@
...
-import { getConfig } from 'import-sort-config';
-import { dirname, extname } from 'path';
+import { DEFAULT_CONFIGS, getConfig } from 'import-sort-config';
 import { TextDocument, window } from 'vscode';
-import onSave from './on-save';
+import { dirname, extname } from 'path';
 import { getConfiguration, getMaxRange } from './utils';
 
+import importSort from 'import-sort';
+import onSave from './on-save';
```
Comment (amatiasq) [inline src/sort.ts]: Please don't change sorting style of a single file can you restore the sorting for this file for this PR? If you want to change import sorting let's do it on a different PR to keep changes clear.

File: src/sort.ts [hunk 5]
```diff
@@ -26,11 +30,27 @@ export function sort(document: TextDocument): string {
     const directory = dirname(fileName);
 
     let result;
+    const config = { ...DEFAULT_CONFIGS };
+    const defaultSortStyle = getConfiguration<string>('default-sort-style');
+    for (const languages in config) {
```
Comment (amatiasq) [inline src/sort.ts]: I know this should be in linting configuration but can you add a blank line before the `for`?

File: src/sort.ts (hunk 4 above)
Comment (cliffkoh) [reply src/sort.ts]: Totally agree - the extension did it but I will fix it!

File: package.json (hunk 3 above)
Comment (cliffkoh) [reply package.json]: I created it - module did not support scoped repos (e.g. `@something/blah`).

File: src/sort.ts (hunk 5 above)
Comment (cliffkoh) [reply src/sort.ts]: :+1:

**Comment by cliffkoh** [comment]
Made the changes requested, please help check :) 

BTW, I found out bypass sort and save does not work, so I fixed it too... (I doubt it ever worked, unless it was a breaking change in VSCode extensibility APIs).

File: package.json (hunk 3 above)
Comment (amatiasq) [reply package.json]: Nice idea! We can also fix that directly in [`module-compact`](https://github.com/amatiasq/import-sort-style-module-compact) but let's add this one anyway.

File: src/on-save.ts [hunk 6]
```diff
@@ -36,16 +36,18 @@ export default {
...
 
     bypass(action) {
         this.unregister();
         const result = action();
-        this.update();
-        return result;
+        return result.then((res) => {
+            this.update();
+            return res;
+        });
```
Comment (amatiasq) [inline src/on-save.ts]: 👍

**Comment by amatiasq** [comment]
Yeah, it's possible that bypass sort and save never worked after the refactoring. This project needs some test to prevent that to happen again.

This PR fixed so many things, I'm going to release a new minor version now. Thanks, it's a great job!
